contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -3403,3 +3412,64 @@ pltsql_set_target_table_alternative(ParseState *pstate, Node *stmt, CmdType comm
 
 	return setTargetTable(pstate, relation, inh, true, requiredPerms);
 }
+
+/*
+ * Update values and nulls arrays with missing column values if any.
+ * Mainly used for Babelfish catalog tables during restore.
+ */
+static void
+fill_missing_values_in_copyfrom(Relation rel, Datum *values, bool *nulls)
+{
+	Oid relid;
+
+	if (!babelfish_dump_restore || IsBinaryUpgrade)
+		return;
+
+	relid = RelationGetRelid(rel);
+	/*
+	 * Insert new dbid column value in babelfish catalog if dump did
+	 * not provide it.
+	 */
+	if (relid == sysdatabases_oid ||
+		relid == namespace_ext_oid ||
+		relid == bbf_view_def_oid)
+	{
+		int16		dbid = 0;
+		AttrNumber	attnum;
+		const char	*prev_current_user;
+
+		attnum = (AttrNumber) attnameAttNum(rel, "dbid", false);
+		Assert(attnum != InvalidAttrNumber);
+
+		if (!nulls[attnum - 1])
+			return;
+
+		/* Get new DB ID. Need sysadmin to do that. */
+		prev_current_user = GetUserNameFromId(GetUserId(), false);
+		bbf_set_current_user("sysadmin");
+		/* For sysdatabases table we need to generate new dbid for the database we are currently restoring. */
+		if (relid == sysdatabases_oid)
+		{
+			if ((dbid = getAvailDbid()) == InvalidDbid)
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_DATABASE_DEFINITION),
+							errmsg("cannot find an available ID for new database.")));
+		}
+		/*
+		 * For all the other catalog tables which contain dbid column, get dbid using current value of the
+		 * babelfish_db_seq sequence. It is ok to fetch current value of the sequence here since we already
+		 * have generated new dbid while inserting into sysdatabases catalog.
+		 */
+		else
+		{
+			RangeVar	*sequence = makeRangeVarFromNameList(stringToQualifiedNameList("sys.babelfish_db_seq"));
```

Comment:
is this safe? there cannot be any race condition where sequence changed due to multiple get DB id calls? can there be multiple restore db calls in paraellel?